File: adapp-kmp/iosApp/iosApp/views/PerformanceView.swift
import SwiftUI
import Shared

struct PerformanceView: View {
    let viewModel: SharedPerformanceViewModel

    @State private var isEratosthenesRunning: Bool = false
    @State private var cpuAverageElapsedTimePerCycle: Double = 0.0

    @State private var isMemoryRunning = false
    @State private var growthRate: Double = 0.0
    @State private var volatility: Double = 0.0
    @State private var intercept: Double = 0.0
    @State private var memoryAverageElapsedTimePerCycle: Double = 0.0

    @State private var items: [String] = []

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 15) {
                Title(accessibleText: Labels.shared.PERFORMANCE_TITLE)

                VStack(alignment: .leading, spacing: 10) {
                    AccText(accessibleText: Labels.shared.CPU_TEST_SECTION)
                        .font(.system(size: 15, weight: .bold))

                    AccText(accessibleText: Labels.shared.cpuTestStatus(isRunning: isEratosthenesRunning))
                    AccText(accessibleText: Labels.shared.cpuAverageTime(cpuAverageTime: cpuAverageElapsedTimePerCycle))

                    Button(action: { viewModel.runSieveOfEratosthenes() }) {
                        AccText(accessibleText: Labels.shared.RUN_CPU_TEST)
                            .frame(maxWidth: .infinity, minHeight: 50)
                            .background(isEratosthenesRunning ? Color.gray : Color.blue)
                            .foregroundColor(.white)
                            .cornerRadius(5)
                    }
                    .accessibilityIdentifier("sieve_eratosthenes_button")
                    .disabled(isEratosthenesRunning)
                }
                .modifier(CardStyle())
                .task {
                    for await running in viewModel.isEratosthenesRunning {
                        self.isEratosthenesRunning = running as! Bool
                    }
                }
                .task {
                    for await averageElapsedTime in viewModel.cpuAverageElapsedTimePerCycle {
                        self.cpuAverageElapsedTimePerCycle = averageElapsedTime as! Double
                    }
                }

                VStack(alignment: .leading, spacing: 10) {
                    AccText(accessibleText: Labels.shared.MEMORY_TEST_SECTION)
                        .font(.system(size: 15, weight: .bold))

                    AccText(accessibleText: Labels.shared.memoryTestStatus(isRunning: isMemoryRunning))
                    AccText(accessibleText: Labels.shared.growthRate(growthRate: growthRate))
                    AccText(accessibleText: Labels.shared.volatility(volatility: volatility))
                    AccText(accessibleText: Labels.shared.intercept(intercept: intercept))
                    AccText(accessibleText: Labels.shared.memoryAverageTime(memoryAverageTime: memoryAverageElapsedTimePerCycle))

                    Button(action: { viewModel.runMemoryStressTest() }) {
                        AccText(accessibleText: Labels.shared.RUN_MEMORY_TEST)
                            .frame(maxWidth: .infinity, minHeight: 50)
                            .background(isMemoryRunning ? Color.gray : Color.blue)
                            .foregroundColor(.white)
                            .cornerRadius(5)
                    }
                    .accessibilityIdentifier("memory_test_button")
                    .disabled(isMemoryRunning)
                }
                .modifier(CardStyle())
                .task {
                    for await running in viewModel.isMemoryRunning {
                        self.isMemoryRunning = running as! Bool
                    }
                }
                .task {
                    for await growthRate in viewModel.growthRate {
                        self.growthRate = growthRate as! Double
                    }
                }
                .task {
                    for await volatility in viewModel.volatility {
                        self.volatility = volatility as! Double
                    }
                }
                .task {
                    for await intercept in viewModel.intercept {
                        self.intercept = intercept as! Double
                    }
                }
                .task {
                    for await averageElapsedTime in viewModel.memoryAverageElapsedTimePerCycle {
                        self.memoryAverageElapsedTimePerCycle = averageElapsedTime as! Double
                    }
                }

                VStack(alignment: .leading, spacing: 10) {
                    AccText(accessibleText: Labels.shared.UI_TEST_SECTION)
                        .font(.system(size: 15, weight: .bold))

                    KmpViewControllerBridge(makeController: MainViewControllerKt.PerformanceController).frame(width: .infinity, height: 650)
                }
                .modifier(CardStyle())
            }
            .padding(EdgeInsets(top: 40, leading: 20, bottom: 20, trailing: 20))
        }
    }
}

File: adapp-kmp/composeApp/src/androidMain/kotlin/lu/etat/adapp_kmp/views/PerformanceView.kt
package lu.etat.adapp_kmp.views

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTagsAsResourceId
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lu.etat.adapp_kmp.composables.AccText
import lu.etat.adapp_kmp.composables.DataCard
import lu.etat.adapp_kmp.composables.ImagesList
import lu.etat.adapp_kmp.composables.PageTitle
import lu.etat.adapp_kmp.resources.AccessibleText
import lu.etat.adapp_kmp.resources.Labels.CPU_TEST_SECTION
import lu.etat.adapp_kmp.resources.Labels.MEMORY_TEST_SECTION
import lu.etat.adapp_kmp.resources.Labels.PERFORMANCE_TITLE
import lu.etat.adapp_kmp.resources.Labels.RUN_CPU_TEST
import lu.etat.adapp_kmp.resources.Labels.RUN_MEMORY_TEST
import lu.etat.adapp_kmp.resources.Labels.UI_TEST_SECTION
import lu.etat.adapp_kmp.resources.Labels.cpuAverageTime
import lu.etat.adapp_kmp.resources.Labels.cpuTestStatus
import lu.etat.adapp_kmp.resources.Labels.growthRate
import lu.etat.adapp_kmp.resources.Labels.intercept
import lu.etat.adapp_kmp.resources.Labels.memoryAverageTime
import lu.etat.adapp_kmp.resources.Labels.memoryTestStatus
import lu.etat.adapp_kmp.resources.Labels.volatility
import lu.etat.adapp_kmp.viewmodels.SharedPerformanceViewModel
import org.koin.compose.viewmodel.koinViewModel

@Composable
fun PerformanceView() {
    val performanceVM: SharedPerformanceViewModel = koinViewModel()

    val isEratosthenesRunning by performanceVM.isEratosthenesRunning.collectAsStateWithLifecycle()
    val cpuAverageElapsedTimePerCycle by performanceVM.cpuAverageElapsedTimePerCycle.collectAsStateWithLifecycle()

    val isMemoryRunning by performanceVM.isMemoryRunning.collectAsStateWithLifecycle()
    val growthRate by performanceVM.growthRate.collectAsStateWithLifecycle()
    val volatility by performanceVM.volatility.collectAsStateWithLifecycle()
    val intercept by performanceVM.intercept.collectAsStateWithLifecycle()
    val memoryAverageElapsedTimePerCycle by performanceVM.memoryAverageElapsedTimePerCycle.collectAsStateWithLifecycle()

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(20.dp, 40.dp),
        verticalArrangement = Arrangement.spacedBy(15.dp)
    ) {
        PageTitle(PERFORMANCE_TITLE)

        DataCard(CPU_TEST_SECTION) {
            AccText(cpuTestStatus(isEratosthenesRunning))
            AccText(cpuAverageTime(cpuAverageElapsedTimePerCycle))

            RunButton(
                accessibleText = RUN_CPU_TEST,
                automationId = "sieve_eratosthenes_button",
                isLoading = isEratosthenesRunning,
                onClick = { performanceVM.runSieveOfEratosthenes() }
            )
        }

        DataCard(MEMORY_TEST_SECTION) {
            AccText(memoryTestStatus(isMemoryRunning))
            AccText(growthRate(growthRate))
            AccText(volatility(volatility))
            AccText(intercept(intercept))
            AccText(memoryAverageTime(memoryAverageElapsedTimePerCycle))

            RunButton(
                accessibleText = RUN_MEMORY_TEST,
                automationId = "memory_test_button",
                isLoading = isMemoryRunning,
                onClick = { performanceVM.runMemoryStressTest() }
            )
        }

        DataCard(UI_TEST_SECTION) {
            ImagesList()
        }
    }
}

@Composable
fun RunButton(accessibleText: AccessibleText, automationId: String, isLoading: Boolean, onClick: () -> Unit) {
    Button(
        onClick = onClick,
        enabled = !isLoading,
        modifier = Modifier.fillMaxWidth().height(50.dp).semantics { testTagsAsResourceId = true }.testTag(automationId),
        shape = RoundedCornerShape(8.dp)
    ) {
        AccText(accessibleText)
    }
}